"""gRPC Parser æœåŠ¡ç«¯å®ç°

åŸºäº WeKnora é¡¹ç›®çš„å¾®æœåŠ¡æ¶æ„è®¾è®¡ï¼Œæä¾›æ–‡ä»¶è§£æ gRPC æœåŠ¡ã€‚

æ ¸å¿ƒåŠŸèƒ½ï¼š
- ParseFile: è§£ææ–‡ä»¶å¹¶è¿”å›æ–‡æœ¬å†…å®¹
- HealthCheck: å¥åº·æ£€æŸ¥ï¼ˆKubernetes å‹å¥½ï¼‰

æ€§èƒ½ä¼˜åŒ–ï¼š
- OCR å¼•æ“é¢„åŠ è½½ï¼ˆé¿å…é¦–æ¬¡è°ƒç”¨å»¶è¿Ÿï¼‰
- å•ä¾‹è¿›ç¨‹æ± ï¼ˆé¿å…é‡å¤åˆ›å»ºè¿›ç¨‹ï¼‰
- è¯¦ç»†çš„æ€§èƒ½ç›‘æ§æ—¥å¿—
"""

import grpc
from concurrent import futures
import logging
import time
import signal
import uuid
import os
import asyncio
from pathlib import Path
from typing import Optional

import sys
from pathlib import Path

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•å’Œ grpc ç›®å½•åˆ° Python è·¯å¾„
project_root = Path(__file__).parent.parent
grpc_dir = Path(__file__).parent
sys.path.insert(0, str(project_root))
sys.path.insert(0, str(grpc_dir))

from generated import parser_pb2, parser_pb2_grpc
from parsers import create_parser
from grpc_health.v1 import health, health_pb2, health_pb2_grpc

logger = logging.getLogger(__name__)


class ParserServiceServicer(parser_pb2_grpc.ParserServiceServicer):
    """è§£æå™¨ gRPC æœåŠ¡å®ç°"""

    def ParseFile(self, request, context):
        """è§£ææ–‡ä»¶ï¼ˆæ ¸å¿ƒæ¥å£ï¼‰

        Args:
            request: ParseRequest æ¶ˆæ¯
            context: gRPC ä¸Šä¸‹æ–‡

        Returns:
            ParseResponse: è§£æå“åº”ï¼ŒåŒ…å«å†…å®¹å’Œå…ƒæ•°æ®
        """
        # ç”Ÿæˆè¯·æ±‚è¿½è¸ª ID
        request_id = str(uuid.uuid4())
        start_time = time.time()

        logger.info(f"[{request_id}] æ”¶åˆ°è§£æè¯·æ±‚: {request.file_name}")

        try:
            # 1. å‚æ•°éªŒè¯
            if not request.file_content:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("file_content ä¸èƒ½ä¸ºç©º")
                logger.error(f"[{request_id}] å‚æ•°éªŒè¯å¤±è´¥: file_content ä¸ºç©º")
                return parser_pb2.ParseResponse()

            if not request.file_name:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("file_name ä¸èƒ½ä¸ºç©º")
                logger.error(f"[{request_id}] å‚æ•°éªŒè¯å¤±è´¥: file_name ä¸ºç©º")
                return parser_pb2.ParseResponse()

            # 2. ä»æ–‡ä»¶åæ£€æµ‹æ–‡ä»¶æ ¼å¼
            file_format = Path(request.file_name).suffix
            if not file_format:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details(f"æ— æ³•ä»æ–‡ä»¶å {request.file_name} ä¸­è¯†åˆ«æ ¼å¼")
                logger.error(f"[{request_id}] æ–‡ä»¶æ ¼å¼è¯†åˆ«å¤±è´¥: {request.file_name}")
                return parser_pb2.ParseResponse()

            logger.info(f"[{request_id}] æ–‡ä»¶: {request.file_name}, æ ¼å¼: {file_format}, å¤§å°: {len(request.file_content)} bytes")

            # 3. åˆ›å»ºè§£æå™¨
            try:
                parser = create_parser(file_format)
            except ValueError as e:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details(str(e))
                logger.error(f"[{request_id}] ä¸æ”¯æŒçš„æ–‡ä»¶æ ¼å¼: {file_format}")
                return parser_pb2.ParseResponse(error_message=str(e))

            # 4. é…ç½®è§£æé€‰é¡¹
            # TODO: æ”¯æŒ request.optionsï¼ˆOCRã€Caption ç­‰ï¼‰
            # å½“å‰ç‰ˆæœ¬ï¼šä½¿ç”¨é»˜è®¤é…ç½®

            # 5. æ‰§è¡Œè§£æ
            logger.info(f"[{request_id}] å¼€å§‹è§£ææ–‡ä»¶: {request.file_name}")
            parse_start = time.time()

            # è°ƒç”¨å¼‚æ­¥ parse æ–¹æ³•
            content = asyncio.run(parser.parse(request.file_content))

            parse_duration = time.time() - parse_start
            logger.info(f"[{request_id}] è§£æå®Œæˆï¼Œè€—æ—¶ {parse_duration*1000:.2f}ms")

            # 6. æ„é€ å“åº”
            total_duration = (time.time() - start_time) * 1000

            metadata = parser_pb2.ParseMetadata(
                page_count=getattr(parser, 'page_count', 0),
                image_count=getattr(parser, 'image_count', 0),
                table_count=getattr(parser, 'table_count', 0),
                ocr_count=getattr(parser, 'ocr_count', 0),
                caption_count=getattr(parser, 'caption_count', 0),
                parse_time_ms=total_duration,
            )

            logger.info(
                f"[{request_id}] è§£æå®Œæˆ: {request.file_name}, "
                f"è€—æ—¶ {total_duration:.2f}ms, "
                f"é¡µæ•° {metadata.page_count}, "
                f"å›¾åƒ {metadata.image_count}, "
                f"è¡¨æ ¼ {metadata.table_count}, "
                f"OCR {metadata.ocr_count}"
            )

            return parser_pb2.ParseResponse(
                content=content,
                metadata=metadata,
            )

        except Exception as e:
            logger.error(f"[{request_id}] è§£æå¤±è´¥: {request.file_name}", exc_info=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(str(e))
            return parser_pb2.ParseResponse(
                error_message=str(e)
            )

    def HealthCheck(self, request, context):
        """å¥åº·æ£€æŸ¥

        Args:
            request: HealthCheckRequest æ¶ˆæ¯
            context: gRPC ä¸Šä¸‹æ–‡

        Returns:
            HealthCheckResponse: å¥åº·çŠ¶æ€
        """
        logger.debug("æ”¶åˆ°å¥åº·æ£€æŸ¥è¯·æ±‚")
        return parser_pb2.HealthCheckResponse(
            status=parser_pb2.HealthCheckResponse.SERVING
        )


def _preload_ocr_engine():
    """é¢„åŠ è½½ OCR å¼•æ“ï¼ˆä¼˜åŒ–é¦–æ¬¡è°ƒç”¨é€Ÿåº¦ï¼‰

    OCR å¼•æ“åˆå§‹åŒ–éœ€è¦ 15-20 ç§’ï¼Œé¦–æ¬¡è°ƒç”¨ä¼šåŠ è½½æ¨¡å‹ï¼ˆ~500MBï¼‰ã€‚
    é¢„åŠ è½½å¯ä»¥é¿å…é¦–æ¬¡è¯·æ±‚çš„é•¿æ—¶é—´ç­‰å¾…ã€‚
    """
    try:
        logger.info("é¢„åŠ è½½ OCR å¼•æ“...")
        from parsers.ocr_engine import get_ocr_engine
        get_ocr_engine()  # å•ä¾‹æ¨¡å¼ï¼Œé¦–æ¬¡è°ƒç”¨ä¼šåŠ è½½æ¨¡å‹
        logger.info("OCR å¼•æ“åŠ è½½å®Œæˆï¼")
    except Exception as e:
        logger.warning(f"OCR å¼•æ“é¢„åŠ è½½å¤±è´¥ï¼ˆä¸å½±å“æœåŠ¡å¯åŠ¨ï¼‰: {e}")


def serve(port: int = 50051, max_workers: int = 10, preload_ocr: bool = True):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨

    Args:
        port: ç›‘å¬ç«¯å£ï¼Œé»˜è®¤ 50051
        max_workers: æœ€å¤§å·¥ä½œçº¿ç¨‹æ•°ï¼Œé»˜è®¤ 10
        preload_ocr: æ˜¯å¦é¢„åŠ è½½ OCR å¼•æ“ï¼Œé»˜è®¤ True
    """
    # é¢„åŠ è½½ OCR å¼•æ“ï¼ˆå¯é€‰ï¼‰
    if preload_ocr:
        _preload_ocr_engine()

    # åˆ›å»º gRPC æœåŠ¡å™¨
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=max_workers),
        options=[
            ('grpc.max_send_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),  # 50MB
        ]
    )

    # æ³¨å†ŒæœåŠ¡
    parser_pb2_grpc.add_ParserServiceServicer_to_server(
        ParserServiceServicer(), server
    )

    # æ³¨å†Œå¥åº·æ£€æŸ¥æœåŠ¡
    health_servicer = health.HealthServicer()
    health_pb2_grpc.add_HealthServicer_to_server(health_servicer, server)
    health_servicer.set(
        "parser.ParserService",
        health_pb2.HealthCheckResponse.SERVING
    )

    # å¯åŠ¨æœåŠ¡å™¨
    server.add_insecure_port(f'[::]:{port}')
    server.start()

    logger.info(f"ğŸš€ Parser gRPC æœåŠ¡å·²å¯åŠ¨ï¼Œç«¯å£: {port}")
    logger.info(f"   - æœ€å¤§å·¥ä½œçº¿ç¨‹æ•°: {max_workers}")
    logger.info(f"   - æœ€å¤§æ¶ˆæ¯å¤§å°: 50MB")
    logger.info(f"   - OCR å¼•æ“é¢„åŠ è½½: {'å·²å¯ç”¨' if preload_ocr else 'å·²ç¦ç”¨'}")

    # ä¼˜é›…å…³é—­å¤„ç†
    def handle_sigterm(signum, frame):
        logger.info("æ”¶åˆ° SIGTERM ä¿¡å·ï¼Œæ­£åœ¨å…³é—­æœåŠ¡...")
        server.stop(grace=5)

    signal.signal(signal.SIGTERM, handle_sigterm)

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        logger.info("æ”¶åˆ°åœæ­¢ä¿¡å·ï¼ˆCtrl+Cï¼‰ï¼Œæ­£åœ¨å…³é—­æœåŠ¡...")
        server.stop(grace=5)


if __name__ == '__main__':
    # é…ç½®æ—¥å¿—
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(process)d - %(thread)d - %(filename)s:%(lineno)d - %(message)s'
    )

    # ä»ç¯å¢ƒå˜é‡è¯»å–é…ç½®
    port = int(os.getenv('PARSER_GRPC_PORT', '50051'))
    max_workers = int(os.getenv('PARSER_GRPC_MAX_WORKERS', '10'))
    preload_ocr = os.getenv('PARSER_GRPC_PRELOAD_OCR', 'true').lower() == 'true'

    # å¯åŠ¨æœåŠ¡
    serve(port=port, max_workers=max_workers, preload_ocr=preload_ocr)
